### PR TITLE
NN-3033 Group prisoner cell history by location and add moved by column

### PR DESCRIPTION
**Pull request by jsrobertson**
image

### CODE REVIEW

**Review by Nimphal — APPROVED** [review]
LGTM!

File: backend/controllers/prisonerProfile/prisonerCellHistory.js [hunk 1]
```diff
@@ -42,27 +44,43 @@ module.exports = ({ oauthApi, prisonApi, logError, page = 0 }) => async (req, re
...
-            ? moment(cell.assignmentEndDateTime).format('YYYY-MM-DDTHH:mm:ss')
-            : today,
-          livingUnitId: cell.livingUnitId,
-          agencyId: cell.agencyId,
-        }
-      })
+    const cellData = await Promise.all(
+      cells
+        .sort((left, right) => sortByDateTime(right.assignmentDateTime, left.assignmentDateTime))
+        .filter(cell => cell.assignmentEndDateTime)
```
Comment (JamieDraperUK) [inline backend/controllers/prisonerProfile/prisonerCellHistory.js:50]: Should this filter be before the sort?

File: backend/controllers/prisonerProfile/prisonerCellHistory.js [hunk 2]
```diff
@@ -42,27 +44,43 @@ module.exports = ({ oauthApi, prisonApi, logError, page = 0 }) => async (req, re
...
+            movedBy: formatName(staffDetails.firstName, staffDetails.lastName),
+          }
+        })
+    )
 
     return res.render('prisonerProfile/prisonerCellHistory.njk', {
-      cellData,
+      cellHistoryGroupedByAgency: hasLength(cellData)
+        ? Object.entries(groupBy(cellData, 'establishment')).map(([key, value]) => {
+            const fromDateString = value[0].movedIn.split(' -')[0]
```
Comment (JamieDraperUK) [inline backend/controllers/prisonerProfile/prisonerCellHistory.js:74]: This is potentially a bit fragile if the date format changes, might be worth using moment to format the date time to a date.

File: backend/controllers/prisonerProfile/prisonerCellHistory.js (hunk 1 above)
Comment (jsrobertson) [reply backend/controllers/prisonerProfile/prisonerCellHistory.js:50]: Yeah makes sense, didn't spot that. 👍

File: backend/tests/prisonerCellHistory.test.js [hunk 3]
```diff
@@ -82,16 +87,23 @@ describe('Prisoner cell history', () => {
       expect(res.render).toHaveBeenCalledWith(
         'prisonerProfile/prisonerCellHistory.njk',
         expect.objectContaining({
-          cellData: [
+          cellHistoryGroupedByAgency: [
```
Comment (JamieDraperUK) [inline backend/tests/prisonerCellHistory.test.js:90]: Be good to get some coverage of grouping across multiple agencies.

File: backend/controllers/prisonerProfile/prisonerCellHistory.js (hunk 2 above)
Comment (jsrobertson) [reply backend/controllers/prisonerProfile/prisonerCellHistory.js:74]: Agreed.

**Review by JamieDraperUK — APPROVED** [review]
Nice work :+1:

**Pull request merged**